src/webauthn/proto/raw_message.rs

```diff
@@ -45,14 +58,525 @@ pub struct Packed {
 }
 
 #[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
 pub struct TPM {
+    pub ver: serde_cbor::Value,
     pub alg: i64,
     #[serde(with = "serde_bytes")]
     pub sig: Vec<u8>,
     #[serde(skip_serializing_if = "Option::is_none")]
     pub x5c: Option<serde_cbor::Value>,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub ecdaa_key_id: Option<serde_cbor::Value>,
+    #[serde(with = "serde_bytes")]
+    pub cert_info: Vec<u8>,
+    #[serde(with = "serde_bytes")]
+    pub pub_area: Vec<u8>,
+}
+
+impl TPM {
+    pub fn verify_structure(&self) -> Result<CertInfo, Error> {
+        if CoseAlgorithmIdentifier::from(self.alg) == CoseAlgorithmIdentifier::NotSupported {
+            return Err(Error::TpmError(TpmError::AlgorithmNotSupported));
+        }
+
+        if let Value::Text(ver) = &self.ver {
+            if ver != "2.0" {
+                return Err(Error::TpmError(TpmError::AttestationVersionNotSupported));
+            }
+        } else {
+            return Err(Error::TpmError(TpmError::AttestationVersionNotSupported));
+        }
+
+        if self.x5c.is_none() {
+            return Err(Error::TpmError(TpmError::CertificateMissing));
+        }
+
+        let cert_info = CertInfo::from_buf(self.cert_info.as_slice())?;
+
+        if cert_info.magic != TPM_GENERATED_VALUE {
+            return Err(Error::TpmError(TpmError::MagicInvalid));
+        }
+
+        Ok(cert_info)
+    }
+
+    pub fn verify_attest(&self, cert_info: &CertInfo, name_alg: TpmAlgId) -> Result<(), Error> {
+        match cert_info.attestation_type {
+            AttestationType::TpmStAttestCertify => {
+                let (name, _) = &cert_info.attested_name;
+                let hash_name = match name_alg {
+                    TpmAlgId::SHA256 => {
+                        let mut hash_name: Vec<u8> = vec![0, 11];
+                        let mut hasher = Sha256::new();
+                        hasher.update(self.pub_area.as_slice());
+                        let mut pub_area_hash = hasher.finalize().to_vec();
+                        hash_name.append(&mut pub_area_hash);
+                        hash_name
+                    }
+                    _ => return Err(Error::TpmError(TpmError::PubAreaHashUnknown(name_alg.into()))),
+                };
+
+                if hash_name != name.to_vec() {
+                    return Err(Error::TpmError(TpmError::AttestedNamePubAreaMismatch));
+                }
+            }
+            _ => return Err(Error::TpmError(TpmError::AttestationTypeInvalid)),
+        }
+
+        Ok(())
+    }
+
+    pub fn verify_extra_data(&self, auth_data: &[u8], client_data_hash: &[u8], extra_data: Option<Vec<u8>>) -> Result<(), Error> {
+        let mut buf = auth_data.to_vec();
+        buf.extend_from_slice(client_data_hash);
```

Comment:
```suggestion
        let mut buf = Vec::with_capacity(auth_data.len() + client_data_hash.len());
        buf.extend_from_slice(auth_data);
        buf.extend_from_slice(client_data_hash);
```
On peut éviter une ré-allocation pour agrandir le vecteur en créant initialement avec la bonne capacitée connue.